---
title: "GenericIntegrationClusteringAndAnalysis"
author: "Sam Darko"
output: html_document
date: "`r Sys.Date()`"

params:
  workingDirectory: "/Users/sam/Documents/portfolio/datasets/single_cell/"
  inputFileString: "cleanedSamples.rds"
  subsetString: "CD8-T"
  
---

```{r setup, include=TRUE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(ggplot2)
library(cowplot)
library(patchwork)
library(dplyr)
library(knitr)
library(r2r)
library(vegan)
library(harmony)

```

## Parameters

Listing parameters definined at the beginning of the rmarkdown document

```{r Parameters, echo = FALSE}

for (name in names(params)) {
    print(paste(name, params[[name]], sep = " -> "))
}

```

## Functions

Functions that are used down the line in this script

A function to compute the priciple components to use in an unbiased manner

```{r ComputePCsToUse, echo = TRUE, warning = FALSE, message = FALSE}
ComputePCsToUseFunction <- function(this.seurat){
  
  ## Find optimum number of PCs to use downstream.  Code is adapted from:
  ## https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html
  
  # Determine percent of variation associated with each PC
  pct <- this.seurat[["pca"]]@stdev / sum(this.seurat[["pca"]]@stdev) * 100
  
  # Calculate cumulative percents for each PC
  cumu <- cumsum(pct)
  
  # Determine which PC exhibits cumulative percent greater than 90% and % variation associated with PC as less than 5
  co1 <- which(cumu > 90 & pct < 5)[1]
  
  # Determine the difference between variation of PC and subsequent PC
  co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
  
  # last point where change of % of variation is more than 0.1%
  pcs <- min(co1, co2)
  
  return(pcs)
}
```

A function to remove genes so they're not used for clustering cells 
For justification: https://pubmed.ncbi.nlm.nih.gov/36473726/

```{r QuietUnwantedGenes, echo = TRUE}

## adapted from:
## https://rdrr.io/github/ncborcherding/Trex/src/R/quietTCRgenes.R

## Define "not in"
`%ni%` <- Negate(`%in%`)

quietUnwantedGenesFunction <- function(this.seurat, unwanted_genes){
  if(inherits(x=this.seurat, what = "Seurat")){
    ## grep variable features (genes) against our unwanted genes list
    this.unwanted_genes <- grep(pattern = unwanted_genes, x = this.seurat[["SCT"]]@var.features, value = TRUE)
    
    ## remove unwanted genes from var.features
    this.seurat[["SCT"]]@var.features <- this.seurat[["SCT"]]@var.features[this.seurat[["SCT"]]@var.features %ni% this.unwanted_genes]
    
    ## remove unwanted genes from scaled matrix produced by SCTransform
    ## probably simpler to just do it in place, I think this looks cleaner
    temp.matrix <- this.seurat[["SCT"]]@scale.data
    temp.matrix <- temp.matrix[!(row.names(temp.matrix) %in% this.unwanted_genes),]
    this.seurat[["SCT"]]@scale.data <- temp.matrix
  
    } else {
        ## working with a vector of genes
        ## grep variable features against unwanted genes
        this.unwanted_genes <- grep(pattern = unwanted_genes, x = this.seurat, value = TRUE)
        this.seurat <- this.seurat[this.seurat %ni% this.unwanted_genes]
    }
  return(this.seurat)
}

```

## Building our Seurat object

```{r ReadInSeuratAndSubset, warning = FALSE}
## read in Seurat object from upstream script
my.seurat <- readRDS(paste(params$workingDirectory, params$inputFileString, sep = ""))

## set the ident to use for subsetting
my.seurat <- SetIdent(my.seurat, value = my.seurat$predicted.celltype.l1)

## subset for level 1 Azimuth designation
my.seurat <- subset(my.seurat, idents = params$subsetString)

## give us a summary of the seurat object
my.seurat

```

## How do things look prior to integration?

We're going to do this as closely as possible to how we do the actual integration  
  
However, we're not going to:  
1. Remove TCR/BCR germline genes from the variable features  
2. Use a rational number of PCs  
3. Actually run Harmony  

```{r NoIntegrationVisualization, warning = FALSE, message = FALSE}
## to throw away later
my.bad.seurat <- my.seurat

## set identity to use, in this case it's the original ID
my.bad.seurat <- SetIdent(my.bad.seurat, value = my.bad.seurat$orig.ident)

## split the dataset into a list by the orignal ID
my.bad.seurat.list <- SplitObject(my.bad.seurat, split.by = "orig.ident")

## normalize and identify variable features for each dataset independently
## it's slightly better to determine variable genes for each sample 
## and then make a final selection later

my.bad.seurat.list <- lapply(X = my.bad.seurat.list, FUN = function(x) {
  x <- SCTransform(x, vst.flavor = "v2", verbose = FALSE, conserve.memory = FALSE, ncells = 5000)
})

## get a list of genes to be used for integration
features <- SelectIntegrationFeatures(object.list = my.bad.seurat.list, nfeatures = 3000)

## merge individual seurat objects
my.bad.seurat <- merge(my.bad.seurat.list[[1]], y = my.bad.seurat.list[2:length(my.bad.seurat.list)], merge.data = TRUE)

## set variable genes from the features selected earlier
VariableFeatures(my.bad.seurat) <- features

## now we can prep SCT markers
my.bad.seurat <- PrepSCTFindMarkers(my.bad.seurat, assay = "SCT", verbose = TRUE)

## run PCA
my.bad.seurat <- RunPCA(object = my.bad.seurat, assay = "SCT", npcs = 50)

#######
## run initial umap
my.bad.seurat <- RunUMAP(object = my.bad.seurat, assay = "SCT", reduction.name = "umap.unintegrated", dims = 1:30)

## Find neighbors using PCs determined earlier
my.bad.seurat <- FindNeighbors(object = my.bad.seurat, assay = "SCT", dims = 1:30)

## Find clusters using neighbors and PCs determined earlier
my.bad.seurat <- FindClusters(object = my.bad.seurat, resolution = 1, algorithm = 1)

## View clusters via UMAP
p1 <- DimPlot(my.bad.seurat, reduction = "umap.unintegrated", label = TRUE, label.size = 8) + NoLegend()
p1

## Let's see how our clusters look with the level 2 Azimuth predictions overlaid
my.bad.seurat <- SetIdent(my.bad.seurat, value = my.bad.seurat$predicted.celltype.l2)
p1 <- DimPlot(my.bad.seurat, reduction = "umap.unintegrated", label = TRUE, label.size = 8, repel = TRUE) + NoLegend()
p1

## see how samples cluster on the UMAP
## it's not good that some clusters are sample specific
my.bad.seurat <- SetIdent(my.bad.seurat, value = "orig.ident")
p2 <- DimPlot(my.bad.seurat, reduction = "umap.unintegrated", label = FALSE)
p2
```


## Diversity measurements without integration

We can look at how well the integration went by looking at cluster and sample diversity
It's just a non-visual way to confirm that no clusters are sample-specific, which would indicate the that those clusters are artifacts and that our integration didn't go well  

In a good integration, all samples would be represented in all clusters and vice versa  

This is not a good integration  

```{r NonIntegratedDiversity}
## generate table for cluster diversity
my.df <- as.data.frame.matrix(table(my.bad.seurat$seurat_clusters, my.bad.seurat$orig.ident))

## calculate cluster diversity and add to a new dataframe
shannon_entropy <- diversity(my.df)
richness <- specnumber(my.df)
evenness <- shannon_entropy/(log(richness))
diversity.df <- data.frame(shannon_entropy, richness, evenness)

## major clusters should have high diversity
## confirming that integration went well
kable(diversity.df)

## transpose data to look at diversity from the sample POV
my.df <- t(my.df)

## calculate sample diversity and add to a new dataframe
shannon_entropy <- diversity(my.df)
richness <- specnumber(my.df)
evenness <- shannon_entropy/(log(richness))
diversity.df <- data.frame(shannon_entropy, richness, evenness)

## samples have similar diversity, which is furthur indication that 
## integration went pretty well
kable(diversity.df)

rm(my.bad.seurat)
```

## Proper Integration

We can see from the previous portion that simply combining the Seurat objects gives poor results  

This is despite the fact that we're only combining cells of the same identity from Azimuth  


## Normalizing and Finding Features

This is the first series of operation in proper integration

```{r IntegrationPrep, message = FALSE, warning = FALSE}

## set identity to use, in this case it's the original ID
my.seurat <- SetIdent(my.seurat, value = my.seurat$orig.ident)

## split the dataset into a list by the orignal ID
my.seurat.list <- SplitObject(my.seurat, split.by = "orig.ident")

## normalize and identify variable features for each dataset independently
## it's slightly better to determine variable genes for each sample 
## and then make a final selection later

my.seurat.list <- lapply(X = my.seurat.list, FUN = function(x) {
  x <- SCTransform(x, vst.flavor = "v2", verbose = FALSE, conserve.memory = FALSE, ncells = 5000)
})

## to remove TCR genes
my.unwanted_genes <- "^TRBV|^TRBD|^TRBJ|^TRBC|^TRDV|^TRDD|^TRDJ|^TRDC|^TRAV|^TRAJ|^TRAC|^TRGV|^TRGJ|^TRGC"

my.seurat.list <- lapply(X = my.seurat.list, FUN = function(x) {
  x <- quietUnwantedGenesFunction(x, my.unwanted_genes)
})

## to remove BCR genes
my.unwanted_genes <- "^IGHC|^IGHD|^IGHJ|^IGHV|^IGKC|^IGKJ|^IGKV|^IGLC|^IGLJ|^IGLV|^IGHA|^IGHD|^IGHE|^IGHG|^IGHM|^IGKC"

my.seurat.list <- lapply(X = my.seurat.list, FUN = function(x) {
  x <- quietUnwantedGenesFunction(x, my.unwanted_genes)
})

## get a list of genes to be used for integration
features <- SelectIntegrationFeatures(object.list = my.seurat.list, nfeatures = 3000)

## merge individual seurat objects
my.seurat <- merge(my.seurat.list[[1]], y = my.seurat.list[2:length(my.seurat.list)], merge.data = TRUE)

## set variable genes from the features selected earlier
VariableFeatures(my.seurat) <- features

## now we can prep SCT markers
my.seurat <- PrepSCTFindMarkers(my.seurat, assay = "SCT", verbose = TRUE)

```

## Running Harmony for integration

This is where we're going to run Harmony for proper integration

```{r HarmonyIntegration, message = FALSE, warning = FALSE}
## adapted from:
## https://github.com/immunogenomics/harmony/issues/41

rm(my.seurat.list)

## Generate PCA using variable genes
my.seurat <- RunPCA(object = my.seurat, assay = "SCT", npcs = 50)

## compute PCs to use in a non-arbitrary fashion
pcs <- ComputePCsToUseFunction(my.seurat)

## probably don't need to run 30 iterations, but let's just be careful
## at the expense of slightly longer run time

my.seurat <- RunHarmony(object = my.seurat,
                        assay.use = "SCT",
                        reduction.use = "pca",
                        dims.use = 1:pcs,
                        group.by.vars = "orig.ident",
                        max_iter = 30,
                        plot_convergence = TRUE)
```

## Visualizing Clustering and Integration 

```{r ClusterVisualization, warning = FALSE, message = FALSE}

## run initial umap
my.seurat <- RunUMAP(object = my.seurat, assay = "SCT", reduction = "harmony", dims = 1:pcs)

## Find neighbors using PCs determined earlier
my.seurat <- FindNeighbors(object = my.seurat, assay = "SCT", reduction = "harmony", dims = 1:pcs)

## Find clusters using neighbors and PCs determined earlier
my.seurat <- FindClusters(object = my.seurat, resolution = 1, algorithm = 1)

## View clusters via UMAP
p1 <- DimPlot(my.seurat, reduction = "umap", label = TRUE, label.size = 8) + NoLegend()
p1

## Let's see how our clusters look with the level 2 Azimuth predictions overlaid
my.seurat <- SetIdent(my.seurat, value = my.seurat$predicted.celltype.l2)
p1 <- DimPlot(my.seurat, reduction = "umap", label = TRUE, label.size = 8, repel = TRUE) + NoLegend()
p1

## see how samples cluster on the UMAP
my.seurat <- SetIdent(my.seurat, value = "orig.ident")
p2 <- DimPlot(my.seurat, reduction = "umap", label = FALSE)

## Do we see any samples segregate? If our integration was good then we shouldn't
p2

## considering that "PBMC10K" is 5' and now 3' GEX like the other two dataset, that's not too bad
```


## Diversity measurements

We can look at how well the integration went by looking at cluster and sample diversity
It's just a non-visual way to confirm that no clusters are sample-specific, which would indicate the that those clusters are artifacts and that our integration didn't go well

In a good integration, all samples would be represented in all clusters and vice versa

```{r Diversity}
## generate table for cluster diversity
my.df <- as.data.frame.matrix(table(my.seurat$seurat_clusters, my.seurat$orig.ident))

## calculate cluster diversity and add to a new dataframe
shannon_entropy <- diversity(my.df)
richness <- specnumber(my.df)
evenness <- shannon_entropy/(log(richness))
diversity.df <- data.frame(shannon_entropy, richness, evenness)

## major clusters should have high diversity
## confirming that integration went well
kable(diversity.df)

## transpose data to look at diversity from the sample POV
my.df <- t(my.df)

## calculate sample diversity and add to a new dataframe
shannon_entropy <- diversity(my.df)
richness <- specnumber(my.df)
evenness <- shannon_entropy/(log(richness))
diversity.df <- data.frame(shannon_entropy, richness, evenness)

## samples have similar diversity, which is furthur indication that 
## integration went pretty well
kable(diversity.df)

```

## Saving cell-type specific data

```{r WriteData}

## and let's write the subsetted object
my.filename <- paste(params$workingDirectory, params$subsetString, ".rds", sep = "")
saveRDS(my.seurat, my.filename)
```

## Session Info
```{r SessionInfo, echo = FALSE}
sessionInfo()
```